sarracen/interpolate.py

```diff
@@ -302,3 +302,120 @@ def _fast_interpolate3d_cross(xterm, yterm, zterm, wfunc, zslice, kernrad, targe
+    :param int_samples: The number of sample points to take when approximating the 2D column kernel.
+    :return: The output image, in a 2-dimensional numpy array.
+    """
+    if pixwidthx <= 0:
+        raise ValueError("pixwidthx must be greater than zero!")
+    if pixwidthy <= 0:
+        raise ValueError("pixwidthy must be greater than zero!")
+    if pixcountx <= 0:
+        raise ValueError("pixcountx must be greater than zero!")
+    if pixcounty <= 0:
+        raise ValueError("pixcounty must be greater than zero!")
+
+    return _fast_3d_interpolate(data[x].to_numpy(),
+                                data[y].to_numpy(),
+                                data[target].to_numpy(),
+                                _get_column_kernel(kernel, int_samples),
+                                int_samples,
+                                kernel.get_radius(),
+                                data['m'].to_numpy(),
+                                data['rho'].to_numpy(),
+                                data['h'].to_numpy(),
+                                xmin,
+                                ymin,
+                                pixwidthx,
+                                pixwidthy,
+                                pixcountx,
+                                pixcounty)
+
+
+# Underlying numba-compiled code for 3D column interpolation.
+@numba.jit(nopython=True, parallel=True, fastmath=True)
+def _fast_3d_interpolate(xparts, yparts, target, wfuncint, int_samples, wrad, mass, rho, h, xmin, ymin,
+                         pixwidthx, pixwidthy, pixcountx, pixcounty):
+    image = np.zeros((pixcounty, pixcountx))
+
+    term = target * mass / (rho * h ** 2)
+
+    # determine maximum and minimum pixels that each particle contributes to
+    ipixmin = np.rint((xparts - wrad * h - xmin) / pixwidthx) \
+        .clip(a_min=0, a_max=pixcountx)
+    jpixmin = np.rint((yparts - wrad * h - ymin) / pixwidthy) \
+        .clip(a_min=0, a_max=pixcounty)
+    ipixmax = np.rint((xparts + wrad * h - xmin) / pixwidthx) \
+        .clip(a_min=0, a_max=pixcountx)
+    jpixmax = np.rint((yparts + wrad * h - ymin) / pixwidthy) \
+        .clip(a_min=0, a_max=pixcounty)
+
+    # iterate through the indexes of non-filtered particles
+    for i in prange(len(term)):
+        # precalculate differences in the x-direction (optimization)
+        dx2i = ((xmin + (np.arange(int(ipixmin[i]), int(ipixmax[i])) + 0.5) * pixwidthx - xparts[i]) ** 2) \
+               * (1 / (h[i] ** 2))
+
+        # determine differences in the y-direction
+        ypix = ymin + (np.arange(int(jpixmin[i]), int(jpixmax[i])) + 0.5) * pixwidthy
+        dy = ypix - yparts[i]
+        dy2 = dy * dy * (1 / (h[i] ** 2))
+
+        # calculate contributions at pixels i, j due to particle at x, y
+        q2 = dx2i + dy2.reshape(len(dy2), 1)
+        wab = np.interp(np.sqrt(q2), np.linspace(0, wrad, int_samples), wfuncint)
+
```

Comment:
This seems like a good way of doing this. You only integrate the kernel once and then use interpolation.